### PR TITLE
[UR][CUDA][L0][HIP] Add virtual memory adapter implementations

### PR DESCRIPTION
**Pull request by steffenlarsen**
This commit adds the adapter implementations of the virtual memory extension functionality to be used in
https://github.com/intel/llvm/pull/8954.

### CODE REVIEW

**Review by jandres742 — APPROVED** [review]

**Comment by jandres742** [comment]
thanks @steffenlarsen !

File: source/adapters/hip/virtual_mem.cpp
Comment (kbenzie) [inline source/adapters/hip/virtual_mem.cpp:69]: nit: Missing trailing new-line.

**Comment by JackAKirk** [comment]
From my understanding this should follow the experimental feature guidelines from https://oneapi-src.github.io/unified-runtime/core/CONTRIB.html ?

Because e.g. 
*
Before making a contribution to the specification you should determine if the change should be made directly to the core specification or introduced as an experimental feature. The criteria we use to make this distinction are as follows:

    The feature exists to enable an experimental feature in a parallel language runtime being built on top of Unified Runtime.

///
*

Do you know if HIP adapter can support this in the future?

**Comment by steffenlarsen** [comment]
> From my understanding this should follow the experimental feature guidelines from https://oneapi-src.github.io/unified-runtime/core/CONTRIB.html ?

My understanding was that this applied to UR interfaces. This does not add any new UR interfaces, just the implementations of some that were added a while ago. We could argue that it should be experimental, but that would mean changing the existing interfaces and I don't know if that is allowed?

> btw Do you know if HIP adapter can support this in the future?

I believe https://docs.amd.com/projects/HIP/en/docs-5.4.1/doxygen/html/group___virtual.html describes the HIP interfaces. They should be similar to the CUDA ones.

**Comment by JackAKirk** [comment]
> > From my understanding this should follow the experimental feature guidelines from https://oneapi-src.github.io/unified-runtime/core/CONTRIB.html ?
> 
> My understanding was that this applied to UR interfaces. This does not add any new UR interfaces, just the implementations of some that were added a while ago. We could argue that it should be experimental, but that would mean changing the existing interfaces and I don't know if that is allowed?
> 
> > btw Do you know if HIP adapter can support this in the future?
> 
> I believe https://docs.amd.com/projects/HIP/en/docs-5.4.1/doxygen/html/group___virtual.html describes the HIP interfaces. They should be similar to the CUDA ones.

I see, makes sense, thanks.

File: source/adapters/level_zero/virtual_mem.cpp
Comment (mmoadeli) [inline source/adapters/level_zero/virtual_mem.cpp:25]: We use an ... ->  We use a ....

**Comment by AllanZyne** [comment]
Hi @steffenlarsen.

I think the DDI tables in `source/adapters/level_zero/ur_interface_loader.cpp` needs to be updated too, like `urGetVirtualMemProcAddrTable` and `urGetPhysicalMemProcAddrTable `.

Thanks.

**Comment by steffenlarsen** [comment]
> Hi @steffenlarsen.
> 
> I think the DDI tables in `source/adapters/level_zero/ur_interface_loader.cpp` need to be updated too, like `urGetVirtualMemProcAddrTable` and `urGetPhysicalMemProcAddrTable `.
> 
> Thanks.

Thank you, @AllanZyne ! It has been addressed and interfaces for the OpenCL and native CPU adapters have been added similar to HIP.

File: source/adapters/hip/physical_mem.cpp
Comment (steffenlarsen) [reply source/adapters/hip/physical_mem.cpp]: They have been removed. 👍

**Comment by cdai2** [comment]
@steffenlarsen   Is there a ETA to land this PR?

**Comment by steffenlarsen** [comment]
@cdai2 - Apologies. I had not noticed the conflict. It has been resolved and now I am waiting for approval and CI. I am hopeful this is ready.

**Comment by steffenlarsen** [comment]
@kbenzie - It looks like the CUDA testing is running into an unrecoverable state, potentially due to overallocation. The OpenCL errors I don't understand. Is there some missing check for support of the API somewhere?

**Comment by pbalcer** [comment]
> @kbenzie - It looks like the CUDA testing is running into an unrecoverable state, potentially due to overallocation. The OpenCL errors I don't understand. Is there some missing check for support of the API somewhere?

This PR seems to be reducing the number of failures. There are no new ones as far as I can tell. So, if that reducing in failures is expected, all you need to do is update the match files.

**Comment by steffenlarsen** [comment]
Thank you, @pbalcer ! I am not too aware of how the current testing is structured for UR and what match file you are referring to. Do you have some documentation to help me understand it better?

**Comment by pbalcer** [comment]
> Thank you, @pbalcer ! I am not too aware of how the current testing is structured for UR and what match file you are referring to. Do you have some documentation to help me understand it better?

There's unfortunately very little in the way of documentation for our test framework. It's something we should add.

Right now CTS has a number of known failing tests on each adapter implementation. The work to improve that is ongoing, and, in the meantime, we've created a list of known failures (the match file) that needs to be updated whenever fixes are made. This is to keep track of progress.

Your PR is fixing a number of conformance tests. For OpenCL, when you scroll down in the test log, you will see an output line saying `Line 7 does not match`. This means that the output of the gtest is different than what we expect it to be. Below that are "Input Lines", which shows you the list of tests that failed. Right after that are `Match Lines`, these are the tests that we expect to fail. In the case of this PR, a few tests that previously were failing are now passing. Which is good. What you need to do now is to update the OpenCL [match file](https://github.com/oneapi-src/unified-runtime/blob/adapters/test/conformance/virtual_memory/virtual_memory_adapter_opencl.match) for virtual memory to reflect that:
```
urVirtualMemGetInfoTestWithParam.Success/Intel_R__OpenCL___{{.*}}___UR_VIRTUAL_MEM_INFO_ACCESS_MODE
urVirtualMemGetInfoTest.InvalidNullHandleContext/Intel_R__OpenCL___{{.*}}_
urVirtualMemGetInfoTest.InvalidNullPointerStart/Intel_R__OpenCL___{{.*}}_
urVirtualMemGetInfoTest.InvalidEnumerationInfo/Intel_R__OpenCL___{{.*}}_
urVirtualMemGranularityGetInfoTest.Success/Intel_R__OpenCL___{{.*}}___UR_VIRTUAL_MEM_GRANULARITY_INFO_MINIMUM
urVirtualMemGranularityGetInfoTest.Success/Intel_R__OpenCL___{{.*}}___UR_VIRTUAL_MEM_GRANULARITY_INFO_RECOMMENDED
urVirtualMemGranularityGetInfoNegativeTest.InvalidEnumeration/Intel_R__OpenCL___{{.*}}_
urVirtualMemGranularityGetInfoNegativeTest.InvalidSizePropSizeSmall/Intel_R__OpenCL___{{.*}}_
urVirtualMemSetAccessTest.Success/Intel_R__OpenCL___{{.*}}_
urVirtualMemSetAccessTest.InvalidNullHandleContext/Intel_R__OpenCL___{{.*}}_
urVirtualMemSetAccessTest.InvalidNullPointerStart/Intel_R__OpenCL___{{.*}}_
```

Each adapter might fail on a different set of tests, so you might need to update other adapter match files as well. Here's how the match file should look like for CUDA:

```
urPhysicalMemCreateTest.Success/NVIDIA_CUDA_BACKEND___{{.*}}___12345
urPhysicalMemCreateTest.InvalidSize/NVIDIA_CUDA_BACKEND___{{.*}}___1
urPhysicalMemCreateTest.InvalidSize/NVIDIA_CUDA_BACKEND___{{.*}}___2
urPhysicalMemCreateTest.InvalidSize/NVIDIA_CUDA_BACKEND___{{.*}}___3
urPhysicalMemCreateTest.InvalidSize/NVIDIA_CUDA_BACKEND___{{.*}}___7
urPhysicalMemCreateTest.InvalidSize/NVIDIA_CUDA_BACKEND___{{.*}}___12
urPhysicalMemCreateTest.InvalidSize/NVIDIA_CUDA_BACKEND___{{.*}}___44
urPhysicalMemCreateTest.InvalidSize/NVIDIA_CUDA_BACKEND___{{.*}}___1024
urPhysicalMemCreateTest.InvalidSize/NVIDIA_CUDA_BACKEND___{{.*}}___4000
urPhysicalMemCreateTest.InvalidSize/NVIDIA_CUDA_BACKEND___{{.*}}___12345
urVirtualMemFreeTest.InvalidNullPointerStart/NVIDIA_CUDA_BACKEND___{{.*}}_
urVirtualMemGranularityGetInfoNegativeTest.InvalidEnumeration/NVIDIA_CUDA_BACKEND___{{.*}}_
```

**Comment by pbalcer** [comment]
The GCC build on CUDA is now segfaulting: https://github.com/oneapi-src/unified-runtime/actions/runs/7033478414/job/19141529054?pr=939

If this is expected, you can add `{{OPT} Segmentation fault` at the beginning of the match file, and then {{OPT}} next to all other lines. This is not a great solution as such match file will then not fail once a test is fixed, but it will suppress the problem.

**Comment by kbenzie** [comment]
@steffenlarsen I can a have a look into what's going on with the CUDA failures today.

**Comment by kbenzie** [comment]
So for the first failing test running in Debug mode there's more informative output. I also added some printing of the size to allocate and the test is wanting to allocate almost the entirety of my 3060Ti's 8GB of VRAM. This is unreasonable, the test needs changed to allow allocation failure or not attempt such large allocations in the first place. Not looked into the other tests yet as this isn't the only failing one.

```
[ RUN      ] urPhysicalMemCreateTest.Success/NVIDIA_CUDA_BACKEND___NVIDIA_GeForce_RTX_3060_Ti___4000
size: 8388608000

UR CUDA ERROR:
        Value:           2
        Name:            CUDA_ERROR_OUT_OF_MEMORY
        Description:     out of memory
        Function:        urPhysicalMemCreate
        Source Location: /home/benie/Projects/oneapi-src/unified-runtime/source/adapters/cuda/physical_mem.cpp:29

/home/benie/Projects/oneapi-src/unified-runtime/test/conformance/virtual_memory/urPhysicalMemCreate.cpp:26: Failure
Expected equality of these values:
  uur::Result(UR_RESULT_SUCCESS)
    Which is: UR_RESULT_SUCCESS
  uur::Result(urPhysicalMemCreate(context, device, size, nullptr, &physical_mem))
    Which is: UR_RESULT_ERROR_UNKNOWN
[  FAILED  ] [  FAILED  ] urPhysicalMemCreateTest.Success/NVIDIA_CUDA_BACKEND___NVIDIA_GeForce_RTX_3060_Ti___4000, where GetParam() = (NVIDIA_GeForce_RTX_3060_Ti_, 4000) (54 ms)
```

File: source/adapters/cuda/physical_mem.cpp
Comment (kbenzie) [inline source/adapters/cuda/physical_mem.cpp:29]: Looks like we need to special case this error code mapping for `cuMemCreate` to line up with the UR spec.

```suggestion
  switch (auto Result = cuMemCreate(&ResHandle, size, &AllocProps, 0)) {
  case CUDA_ERROR_INVALID_VALUE:
    return UR_RESULT_ERROR_INVALID_SIZE;
  default:
    UR_CHECK_ERROR(Result);
  }
```

**Review by kbenzie — CHANGES_REQUESTED** [review]
There are three bugs in the CTS, since you've not touched those files here's the diff to fix those. These fixes in addition to the error code special case below resulting in 100% pass rate of the virtual memory test suite on the CUDA adapter on my 3060Ti locally.

```diff
diff --git a/test/conformance/virtual_memory/urPhysicalMemCreate.cpp b/test/conformance/virtual_memory/urPhysicalMemCreate.cpp
index 078b0e68..e5124da1 100644
--- a/test/conformance/virtual_memory/urPhysicalMemCreate.cpp
+++ b/test/conformance/virtual_memory/urPhysicalMemCreate.cpp
@@ -16,8 +16,7 @@ struct urPhysicalMemCreateTest
     size_t size;
 };

-UUR_TEST_SUITE_P(urPhysicalMemCreateTest,
-                 ::testing::Values(1, 2, 3, 7, 12, 44, 1024, 4000, 12345),
+UUR_TEST_SUITE_P(urPhysicalMemCreateTest, ::testing::Values(1, 2, 3, 7, 12, 44),
                  uur::deviceTestWithParamPrinter<size_t>);

 TEST_P(urPhysicalMemCreateTest, Success) {
diff --git a/test/conformance/virtual_memory/urVirtualMemFree.cpp b/test/conformance/virtual_memory/urVirtualMemFree.cpp
index 6cb87955..3b4c4bab 100644
--- a/test/conformance/virtual_memory/urVirtualMemFree.cpp
+++ b/test/conformance/virtual_memory/urVirtualMemFree.cpp
@@ -19,5 +19,5 @@ TEST_P(urVirtualMemFreeTest, InvalidNullHandleContext) {

 TEST_P(urVirtualMemFreeTest, InvalidNullPointerStart) {
     ASSERT_EQ_RESULT(urVirtualMemFree(context, nullptr, size),
-                     UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+                     UR_RESULT_ERROR_INVALID_NULL_POINTER);
 }
diff --git a/test/conformance/virtual_memory/urVirtualMemGranularityGetInfo.cpp b/test/conformance/virtual_memory/urVirtualMemGranularityGetInfo.cpp
index d4feccd6..df8c2879 100644
--- a/test/conformance/virtual_memory/urVirtualMemGranularityGetInfo.cpp
+++ b/test/conformance/virtual_memory/urVirtualMemGranularityGetInfo.cpp
@@ -60,7 +60,7 @@ TEST_P(urVirtualMemGranularityGetInfoNegativeTest, InvalidEnumeration) {
                          context, device,
                          UR_VIRTUAL_MEM_GRANULARITY_INFO_FORCE_UINT32, 0,
                          nullptr, &size),
-                     UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+                     UR_RESULT_ERROR_INVALID_ENUMERATION);
 }

 TEST_P(urVirtualMemGranularityGetInfoNegativeTest,
```

**Comment by pbalcer** [comment]
You need to update the match files again ;) cuda's should be empty, don't know how the test changes will after other adapters.

**Review by kbenzie — APPROVED** [review]

**Comment by fabiomestre** [comment]
I have updated the target branch of this PR from the `adapters` branch to the `main` branch.
Development in UR is moving back to `main`. The `adapters` branch will soon be deleted.

**Comment by kbenzie** [comment]
> Could you please help to merge this PR? It has "ready to merge" flag now.
> We need to test and submit more PRs basing on this PR.

Apologies for the slow rate of merging. However we do have a rather large backlog of PR's which are also gated on merging to intel/llvm which are bug fixes for the next release. We'll do our best to move this along as quickly as we can.

**Comment by kbenzie** [comment]
Marking as draft until https://github.com/intel/llvm/pull/8954 passes.

**Comment by aarongreig** [comment]
Even though these changes shouldn't affect the adapter behaviour for SYCL without all the associated pi etc changes in https://github.com/intel/llvm/pull/8954, I've made https://github.com/intel/llvm/pull/12198 just so we have a commit on the intel llvm side to track the adapter changes back to if needed. I'll merge this PR once the testing there passes (should be soon, just waiting for the final job to get picked up by a runner)

**Pull request merged**
